src/content/blog/2024-12-04-differences-between-linting-and-type-checking.md

```diff
@@ -0,0 +1,248 @@
+---
+layout: post
+title: "Differences between linting and type checking"
+teaser: "Linters such as ESLint and type checkers such as TypeScript catch different areas of code defects — and are best used in conjunction with each other."
+authors:
+  - joshuakgoldberg
+categories:
+  - Storytime
+tags:
+  - Linting
+---
+
+Linters and type checkers are two kinds of tools that both analyze code and report on detected issues.
+They may seem similar at first.
+However, the two kinds of tools detect very different issues and are useful in different ways.
+
+## Definitions
+
+*"Static analysis"* is the term for tooling that checks code for issues without running it.
+There are three forms of static analysis commonly used in web development projects today:
+
+1. **Formatters**: only print your code quickly, without worrying about logic
+2. **Linters**: execute individually configurable checks known as "lint rules"
+3. **Type checkers**: collect all files into a full understanding of the project
```

Comment:
I don't think these descriptions are accurate. You're describing how ESLint works and how TypeScript works, but that doesn't necessarily mean all linters and type checkers work like this.

I'd focus instead on linting as the systematic evaluation of patterns in the structure of source code (maybe even briefly describe the original C `lint` tool) while type checking is an evaluation of data flow throughout the program.

Reply:
It may also be worth nothing that compiled languages do type checking as part of the compilation step, but since JavaScript has no such steps, TypeScript is run separately to do this.